File: standalone_app/planner.py
import math
from typing import List, Tuple

import numpy as np
from PIL import Image, ImageDraw

from utils import (
    CIRCLE_RADIUS,
    PlanResult,
    binarize_mask,
    fill_small_holes,
    inscribed_center,
    largest_connected_component,
)


def compute_faz_center(faz_mask: np.ndarray) -> Tuple[int, int]:
    faz_bin = binarize_mask(faz_mask)
    faz_lcc = largest_connected_component(faz_bin)
    return inscribed_center(faz_lcc)


def generate_ring_points(
    center: Tuple[int, int],
    radius: float,
    min_distance: int = 50,
) -> np.ndarray:
    if radius <= 0:
        return np.empty((0, 2), dtype=np.int32)
    if radius * 2 < min_distance:
        return np.empty((0, 2), dtype=np.int32)
    angle_step = 2 * math.asin(min_distance / (2 * radius))
    if angle_step <= 0:
        return np.empty((0, 2), dtype=np.int32)
    num_points = int(math.floor(2 * math.pi / angle_step))
    if num_points < 1:
        return np.empty((0, 2), dtype=np.int32)
    angles = np.linspace(0, 2 * math.pi, num=num_points, endpoint=False)
    cx, cy = center
    xs = cx + radius * np.cos(angles)
    ys = cy + radius * np.sin(angles)
    coords = np.stack([xs, ys], axis=1)
    coords_int = np.rint(coords).astype(np.int32)
    return coords_int


def plan_surgery(
    image: Image.Image,
    mask: np.ndarray,
    faz_center: Tuple[int, int],
    radius_step: int = 50,
) -> PlanResult:
    overlay = image.copy()
    draw = ImageDraw.Draw(overlay)

    mask_bin = binarize_mask(mask)
    mask_bin = fill_small_holes(mask_bin, area_threshold=200)
    h, w = mask_bin.shape
    cx, cy = faz_center
    max_radius = int(
        max(
            math.hypot(cx, cy),
            math.hypot(cx, h - 1 - cy),
            math.hypot(w - 1 - cx, cy),
            math.hypot(w - 1 - cx, h - 1 - cy),
        )
    )

    all_curve_points: List[np.ndarray] = []
    all_centers: List[Tuple[int, int]] = []

    radius = radius_step
    while radius <= max_radius:
        ring_points = generate_ring_points(faz_center, radius, min_distance=50)
        if len(ring_points) > 0:
            all_curve_points.append(ring_points)
        for x, y in ring_points:
            if 0 <= x < w and 0 <= y < h and mask_bin[y, x] > 0:
                all_centers.append((int(x), int(y)))
        radius += radius_step

    for center in all_centers:
        x, y = center
        draw.ellipse(
            (x - CIRCLE_RADIUS, y - CIRCLE_RADIUS, x + CIRCLE_RADIUS, y + CIRCLE_RADIUS),
            outline=(0, 0, 255),
            width=2,
        )

    return PlanResult(overlay=overlay, curve_points=all_curve_points, circle_centers=all_centers)

File: standalone_app_v3_5_go/internal/planner/planner.go
package planner

import (
	"image"
	"image/color"
	"image/draw"
	"math"

	"eyesam/standalone_app_v3_5_go/internal/utils"
)

func ComputeFAZCenter(mask utils.Mask) image.Point {
	if mask.Width == 0 || mask.Height == 0 {
		return image.Point{}
	}
	return image.Point{X: mask.Width / 2, Y: mask.Height / 2}
}

func GenerateRingPoints(center image.Point, radius float64, minDistance int) []image.Point {
	if radius <= 0 || minDistance <= 0 {
		return nil
	}
	if radius*2 < float64(minDistance) {
		return nil
	}
	angleStep := 2 * math.Asin(float64(minDistance)/(2*radius))
	if angleStep <= 0 {
		return nil
	}
	numPoints := int(math.Floor(2 * math.Pi / angleStep))
	if numPoints < 1 {
		return nil
	}
	points := make([]image.Point, 0, numPoints)
	for i := 0; i < numPoints; i++ {
		angle := 2 * math.Pi * float64(i) / float64(numPoints)
		x := center.X + int(math.Round(radius*math.Cos(angle)))
		y := center.Y + int(math.Round(radius*math.Sin(angle)))
		points = append(points, image.Point{X: x, Y: y})
	}
	return points
}

func PlanSurgery(
	img image.Image,
	mask utils.Mask,
	fazCenter image.Point,
	spotDiameter int,
	spotDistance int,
) utils.PlanResult {
	bounds := img.Bounds()
	overlay := image.NewRGBA(bounds)
	draw.Draw(overlay, bounds, img, bounds.Min, draw.Src)

	circleRadius := int(math.Max(1, math.Round(float64(spotDiameter)/2)))
	radiusStep := int(math.Max(1, math.Round(float64(spotDiameter+spotDistance))))
	minDistance := int(math.Max(1, math.Round(float64(spotDiameter+spotDistance))))

	maxRadius := int(math.Max(
		hypot(fazCenter.X, fazCenter.Y),
		math.Max(
			hypot(bounds.Dx()-1-fazCenter.X, fazCenter.Y),
			hypot(fazCenter.X, bounds.Dy()-1-fazCenter.Y),
		),
	))

	var curvePoints [][]image.Point
	var centers []image.Point

	for radius := radiusStep; radius <= maxRadius; radius += radiusStep {
		ringPoints := GenerateRingPoints(fazCenter, float64(radius), minDistance)
		if len(ringPoints) > 0 {
			curvePoints = append(curvePoints, ringPoints)
		}
		for _, pt := range ringPoints {
			if pt.X < 0 || pt.Y < 0 || pt.X >= mask.Width || pt.Y >= mask.Height {
				continue
			}
			if mask.At(pt.X, pt.Y) == 0 {
				continue
			}
			centers = append(centers, pt)
		}
	}

	for _, center := range centers {
		drawCircle(overlay, center, circleRadius, color.RGBA{R: 0, G: 112, B: 255, A: 255})
	}

	return utils.PlanResult{Overlay: overlay, CurvePoints: curvePoints, CircleCenters: centers}
}

func drawCircle(img *image.RGBA, center image.Point, radius int, stroke color.RGBA) {
	if radius <= 0 {
		return
	}
	minX := center.X - radius
	maxX := center.X + radius
	minY := center.Y - radius
	maxY := center.Y + radius
	for y := minY; y <= maxY; y++ {
		for x := minX; x <= maxX; x++ {
			dx := x - center.X
			dy := y - center.Y
			if dx*dx+dy*dy > radius*radius {
				continue
			}
			if x < 0 || y < 0 || x >= img.Bounds().Dx() || y >= img.Bounds().Dy() {
				continue
			}
			img.SetRGBA(x, y, stroke)
		}
	}
}

func hypot(x, y int) float64 {
	return math.Hypot(float64(x), float64(y))
}
